flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java

```diff
@@ -165,8 +171,16 @@ public boolean hasCleanJobResultEntryInternal(JobID jobId) throws IOException {
 
     @Override
     public Set<JobResult> getDirtyResultsInternal() throws IOException {
+        final FileStatus[] statuses = fileSystem.listStatus(this.basePath);
+        if (statuses == null) {
+            LOG.warn(
+                    "The JobResultStore directory '"
+                            + basePath
+                            + "' was deleted. No persisted JobResults could be recovered.");
+            return Collections.emptySet();
```

Comment:
Isn't this expected if you start out in a fresh environment?

Reply:
This was added because I noticed that the `LocalFileSystem.listStatus` is implemented in a way that it returns `null` if the folder doesn't exist. I added this `null` check (instead of fixing it in `LocalFileSystem`) because the contract of the method isn't clear enough and I didn't want to mingle with the public interface `FileSystem`. `HadoopFileSystem.listStatus` would return an empty array in that case (because of the contract of the internally [Hadoop FileSystem](https://hadoop.apache.org/docs/stable/api/org/apache/hadoop/fs/FileSystem.html#listStatus-org.apache.hadoop.fs.Path-)).

Therefore, this guard is to cover cases where a `LocalFileSystem` or any other `FileSystem` implementation is used that allows returning `null` here.

Reply:
It's fine to check for that, but why log a warning?

Reply:
because it's an unexpected state of Flink which might lead to weird behavior. Some FileSystems would complain about a missing directory whereas other filesystem might allow to create a file within this directory anyway (e.g. object stores?)

Reply:
One could also argue that it should throw an `IllegalStateException` and fail fatally. WDYT? 🤔

Reply:
I agree that we could solve the issue by just creating the directory lazily. But this would mean that we would need to check its existence every time a dirty result is created. That might be expensive for an empty directory (considering that we're planning to delete files after the job cleanup finished). Hence, I went for an early failure with a meaningful error message instead. We should be able to expect that the directory is available because it's created in the initialization of the `FileSystemJobResultStore` - returning `null` indicates an inconsistent state and, therefore, should lead to an error.

Returning an empty set isn't an option it might hide the issue only for a bit until Flink tries to access the file in the folder or create new once. In this sense, failing earlier is the better approach in my opinion.